Sources/Purchasing/Purchases/TransactionPoster.swift

```diff
@@ -0,0 +1,306 @@
+    ) {
+        self.productsManager = productsManager
+        self.receiptFetcher = receiptFetcher
+        self.currentUserProvider = currentUserProvider
+        self.attribution = attribution
+        self.backend = backend
+        self.paymentQueueWrapper = paymentQueueWrapper
+        self.systemInfo = systemInfo
+        self.operationDispatcher = operationDispatcher
+    }
+
+    func handlePurchasedTransaction(_ transaction: StoreTransaction,
+                                    presentedOfferingID: String?,
+                                    storefront: StorefrontType?,
+                                    source: PurchaseSource,
+                                    completion: @escaping PurchaseCompletedBlock) {
+        self.receiptFetcher.receiptData(
+            refreshPolicy: self.refreshRequestPolicy(forProductIdentifier: transaction.productIdentifier)
+        ) { receiptData in
+            if let receiptData = receiptData, !receiptData.isEmpty {
+                self.fetchProductsAndPostReceipt(
+                    withTransaction: transaction,
+                    receiptData: receiptData,
+                    source: source,
+                    storefront: storefront,
+                    presentedOfferingID: presentedOfferingID,
+                    completion: completion
+                )
+            } else {
+                self.handleReceiptPost(withTransaction: transaction,
+                                       result: .failure(.missingReceiptFile()),
+                                       subscriberAttributes: nil,
+                                       completion: completion)
+            }
+        }
+    }
+
+    func finishTransactionIfNeeded(
+        _ transaction: StoreTransactionType,
+        completion: @escaping @Sendable @MainActor () -> Void
+    ) {
+        @Sendable
+        func complete() {
+            self.operationDispatcher.dispatchOnMainActor(completion)
+        }
+
+        guard self.finishTransactions else {
+            complete()
+            return
+        }
+
+        Logger.purchase(Strings.purchase.finishing_transaction(transaction))
+
+        transaction.finish(self.paymentQueueWrapper.paymentQueueWrapperType, completion: complete)
+    }
+
+    func markSyncedIfNeeded(
+        subscriberAttributes: SubscriberAttribute.Dictionary?,
+        error: BackendError?
+    ) {
+        if let error = error {
+            guard error.successfullySynced else { return }
+
+            if let attributeErrors = (error as NSError).subscriberAttributesErrors, !attributeErrors.isEmpty {
+                Logger.error(Strings.attribution.subscriber_attributes_error(
+                    errors: attributeErrors
+                ))
+            }
+        }
+
+        self.attribution.markAttributesAsSynced(subscriberAttributes, appUserID: self.appUserID)
+    }
+
+}
+
+// MARK: - Implementation
+
+private extension TransactionPoster {
+
+    /// Called as a result a purchase.
```

Comment:
... or a renewal, right? 

I realize a renewal is a type of purchase, but given that in parts of the codebase we make distinctions between one and the other to determine initiation source, I think it might be good to clarify

Reply:
and it might also be good to clarify that this won't be called from restores, since restores aren't tied to specific transactions

Reply:
This was an old comment but yeah I'll update 👍🏻

Reply:
Actually I'll remove this comment, it's not that relevant in this context anymore.